compiler/src/openxla/compiler/nvgpu/Transforms/ConvertMHLOToCUDNN.cpp

```diff
@@ -23,72 +24,144 @@
+    auto cast_op = rewriter.create<UnrealizedConversionCastOp>(
+        arg.getLoc(), arg.getType(), graph.getArgument(index));
+    mapping.map(arg, cast_op.getResult(0));
   }
-};
+  ValueRange results;
+  for (Operation* op : llvm::reverse(ops)) {
+    results = rewriter.clone(*op, mapping)->getResults();
+    mapping.map(op->getResults(), results);
+  }
+  rewriter.create<ReturnOp>(root->getLoc(), results);
+
+  // Replace root with cudnn.call op.
+  rewriter.setInsertionPoint(root);
+  rewriter.replaceOpWithNewOp<CallOp>(
+      root, result.getType(), graph.getName(),
+      ArrayRef<Value>(arguments.data(), arguments.size()));
+
+  return success();
+}
+
+// Returns the 'min' value of clamp 'op', if it can be converted to relu.
+static FailureOr<llvm::APFloat> matchRelu(stablehlo::ClampOp op,
+                                          PatternRewriter& rewriter) {
+  llvm::APFloat min = llvm::APFloat::IEEEdouble();
+  if (!matchPattern(op.getMin().getDefiningOp(), m_ConstantFloat(&min))) {
+    return rewriter.notifyMatchFailure(op, "expected constant min");
+  }
+  llvm::APFloat max = llvm::APFloat::IEEEdouble();
+  if (!matchPattern(op.getMax().getDefiningOp(), m_ConstantFloat(&max)) ||
+      !max.isNaN()) {
+    return rewriter.notifyMatchFailure(op, "expected NaN max");
+  }
+  return min;
+}
+
+// Matches any clamp 'op' which can be converted to relu and outlines it into
+// a cudnn.graph op.
+// TODO(chsigg): extend this to match a set of ops (determined through a cost
+// model) to run as a single graph.
+static LogicalResult outlineClamp(stablehlo::ClampOp op,
+                                  PatternRewriter& rewriter) {
+  if (failed(matchRelu(op, rewriter))) return failure();
+  return outlineToGraph(op.getResult(), op.getOperand(), rewriter);
+}
+
+// Converts a clamp 'op' into a cudnn.pointwise_relu.
+static LogicalResult convertClamp(stablehlo::ClampOp op,
+                                  PatternRewriter& rewriter) {
+  if (!op->getParentOfType<GraphOp>())
+    return rewriter.notifyMatchFailure(op, "expected child of graph");
+  auto min_or = matchRelu(op, rewriter);
+  if (failed(min_or)) return failure();
+  auto operand = op.getOperand();
+  TensorType tensor_type = getCudnnTensorType(operand.getType());
+  auto cast_op = rewriter.create<UnrealizedConversionCastOp>(
+      operand.getLoc(), tensor_type, operand);
+  rewriter.replaceOpWithNewOp<PointWiseReluOp>(
+      op, tensor_type, cast_op.getResult(0), tensor_type.getElementType(),
+      APFloat(min->convertToDouble()));
+  return success();
+}
+
+namespace {
 
 class ConvertMHLOToCUDNN : public ::impl::ConvertMHLOToCUDNNBase<
     ConvertMHLOToCUDNN> {
  public:
   void runOnOperation() override {
-    RewritePatternSet patterns(&getContext());
-    patterns.insert<ConvertClamp>(&getContext());
+    RewritePatternSet(&getContext()).add(&outlineClamp);
+    // TODO(chsigg): don't outline and convert in the same pattern application.
```

Comment:
How do you want to split it in the future? I'm thinking about 1. greedy pattern rewriter to form cudnn.graph ops 2. dialect conversion to legalize stablehlo inside cudnn graphs into cudnn ops. And run these two transformations inside a single pass.

Reply:
1. For starters, matching specific subgraphs with greedy patterns might be sufficient. Medium term though, outlining subgraphs will become an optimization problem with some cost model, and the simple greedy rewriters might not be enough.

2. Yes, that's what I was thinking too.